File: Python/392. Is Subsequence.py
"""
Given two strings s and t, return true if s is a subsequence of t, or false otherwise.

A subsequence of a string is a new string that is formed from the original string by deleting some (can be none) 
    of the characters without disturbing the relative positions of the remaining characters. 
    (i.e., "ace" is a subsequence of "abcde" while "aec" is not).
"""
# 392. Is Subsequence
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if s == "" or s == t:
            return True
        
        it = iter(t)
        return all(char in it for char in s)
    
# Test Code
solution = Solution()
#s = "abc"
#t = "ahbgdc"
# Expect True

s = "aaaaaa"
t = "bbaaaa"
# Expect False
print(solution.isSubsequence(s, t))

File: Rust/392_Is_Subsequence.rs
/*
Given two strings s and t, return true if s is a subsequence of t, or false otherwise.

A subsequence of a string is a new string that is formed from the original string by deleting some (can be none) 
    of the characters without disturbing the relative positions of the remaining characters. 
    (i.e., "ace" is a subsequence of "abcde" while "aec" is not).
*/
/* 392. Is Subsequence */
struct Solution;
impl Solution {
    pub fn is_subsequence(s: String, t: String) -> bool {
        if (s.is_empty()) { return true; }
        
        let mut s_chars = s.chars();
        let mut s_iter = s_chars.next();
        
        for t_char in t.chars() {
            if Some(t_char) == s_iter {
                s_iter = s_chars.next();
            }
        }
        s_iter.is_none()
    }
}

fn main() {
    //let s = String::from("abc");
    //let t = String::from("ahbgdc");
    // Expect true
    
    let s = String::from("aaaaaa");
    let t = String::from("bbaaaa");
    // Expect false
    
    let result = Solution::is_subsequence(s, t);
    println!("{}", result);
}